Homeworks/Homework_5/part_4.py

```diff
@@ -0,0 +1,130 @@
+import random
+from datetime import datetime, timezone, timedelta
+from Homeworks.Homework_5.utils import establish_db_connection, get_logger, get_currency
+from itertools import groupby
+
+logger = get_logger()
+
+
+@establish_db_connection
+def randomly_assign_discounts(cursor):
+    """
+    Randomly assigns discounts to a specified number of users.
+
+    :param: num_users (int): The number of users to assign discounts to.
+    :return: dict: A dictionary containing user IDs as keys and assigned discounts as values.
+    :raises: ValueError: If the number of users is too high or negative.
+    """
+    num_users = random.randint(1, 10)
+    logger.info(f"Randomly assigning discounts to {num_users} users")
+
+    credits_discount = [25, 30, 50]
+    user_ids = [i[0] for i in cursor.execute('SELECT Id FROM User')]
+    user_ids = random.sample(user_ids, num_users)
+    user_with_discounts = {user_id: random.choice(credits_discount) for user_id in user_ids}
+    logger.info("Discount assignment completed")
+    return user_with_discounts
+
+
+@establish_db_connection
+def get_debtors_full_name(cursor):
+    """
+    Retrieves the full names of users with negative account balances.
+
+    :return: list: A list of full names of users with negative account balances.
+    """
+    logger.info("Retrieving full names of users with negative account balances")
+    debtors_id = [i[0] for i in cursor.execute('SELECT User_id FROM Account WHERE Amount < 0')]
+    return [' '.join(list(cursor.execute(f'SELECT Name, Surname FROM User WHERE Id = {debtor_id}'))[0])
+            for debtor_id in debtors_id]
```

Comment:
I like comprehensions the most. but unfortunately,  in this particular case I would suggest you to simplify it in some way. in order to make expression more readable.